File: ch5/akkademaid-scala/src/test/scala/com/akkademy/AssignActorsToDispatcherTest.scala
package com.akkademy

import akka.actor.{ActorRef, ActorSystem, Props}
import akka.routing._
import com.akkademy.TestHelper.TestCameoActor
import org.scalatest.{FlatSpec, Matchers}

import scala.concurrent.duration._
import scala.concurrent.{Await, Promise}

class AssignActorsToDispatcherTest extends FlatSpec with Matchers {
  val system = ActorSystem()

  "ActorsAssignedToDispatcher" should "do work concurrently" in {

    val p = Promise[String]()

    val actors: IndexedSeq[ActorRef] = (0 to 7).map(x => {
      system.actorOf(Props(classOf[ArticleParseActor]).
        withDispatcher("article-parsing-dispatcher"))
    })

    val workerRouter = system.actorOf(RoundRobinGroup(
      actors.map(x => x.path.toStringWithoutAddress).toList).props(),
      "workerRouter")

    val cameoActor: ActorRef =
      system.actorOf(Props(new TestCameoActor(p)))

    (0 to 2000).foreach(x => {
      workerRouter.tell(
        new ParseArticle(TestHelper.file)
        , cameoActor);
    })

    TestHelper.profile(() => Await.ready(p.future, 20 seconds), "ActorsAssignedToDispatcher")
  }

}

File: ch5/akkademaid-java/src/test/java/com/akkademy/AssignActorsToDispatcherTest.java
package com.akkademy;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.routing.RoundRobinGroup;
import akka.routing.RoundRobinPool;
import akkademy.ArticleParseActor;
import akkademy.ParseArticle;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class AssignActorsToDispatcherTest {
    ActorSystem system = ActorSystem.create();
    @Test
    public void shouldReadFilesWithActorsInAnotherDispatcher() throws Exception {

        List<ActorRef> routees = Arrays.asList(1,2,3,4,5,6,7,8).stream().map(x ->
                system.actorOf(Props.create(ArticleParseActor.class).
                withDispatcher("article-parsing-dispatcher"))
        ).collect(Collectors.toList());

        Iterable<String> routeeAddresses = routees.
                stream().
                map(x -> x.path().toStringWithoutAddress()).
                collect(Collectors.toList());

        ActorRef workerRouter = system.actorOf(new RoundRobinGroup(routeeAddresses).props());

        CompletableFuture future = new CompletableFuture();
        ActorRef cameoActor = system.actorOf(Props.create(TestCameoActor.class, future));

        IntStream.range(0, 2000).forEach(x -> {
                    workerRouter.tell(
                            new ParseArticle(TestHelper.file)
                            , cameoActor);
                }
        );

        long start = System.currentTimeMillis();
        future.get();
        long elapsedTime = System.currentTimeMillis() - start;
        System.out.println("ActorsAssignedToDispatcherTest Took: " + elapsedTime);

    }
}
